### PR TITLE
NLP refactoring - Stage 2

### PR DESCRIPTION
**Pull request by ekmb**
Signed-off-by: Evelina Bakhturina <ebakhturina@nvidia.com>

Stage 2 of NLP refactoring:
+ Cleaning up and restructuring of functions and files in nlp collection.
+ Cleaning up losses
++ Added weighting option to LossAggregatorNM 
++ Moved LossAggregatorNM to the losses.py in the backend common
++ Splited JointIntentSlotLoss into two separate common losses and removed it
++ Merged MaskedLanguageModelingLossNM, PaddedSmoothedCrossEntropyLossNM and SmoothedCrossEntropyLoss into a unified loss SmoothedCrossEntropyLoss
++ Changed QuestionAnsweringLoss to a more general name SpanningLoss
++ Changed TRADEMaskedCrossEntropy to a more general name MaskedXEntropyLoss
++ Removed   TokenClassificationLoss, CrossEntropyLoss3D and  JointIntentSlotLoss
++ Added weighting and masking support to CrossEntropyLossNM
++ Added dynamic port sizes to CrossEntropyLossNM
++ Changed CrossEntropyLoss to CrossEntropyLossNM to prevent confusion with pytorch's CrossEntropyLoss

### CODE REVIEW

**Comment by VahidooX** [comment]
Stage 2 of NLP refactoring:
+ Cleaning up and restructuring of functions and files in nlp collection.
+ Cleaning up losses
++ Added weighting option to LossAggregatorNM 
++ Moved LossAggregatorNM to the losses.py in the backend common
++ Splited JointIntentSlotLoss into two separate common losses and removed it
++ Merged MaskedLanguageModelingLossNM, PaddedSmoothedCrossEntropyLossNM and SmoothedCrossEntropyLoss into a unified loss SmoothedCrossEntropyLoss
++ Changed QuestionAnsweringLoss to a more general name SpanningLoss
++ Changed TRADEMaskedCrossEntropy to a more general name MaskedXEntropyLoss
++ Removed   TokenClassificationLoss, CrossEntropyLoss3D and  JointIntentSlotLoss
++ Added weighting and masking support to CrossEntropyLossNM
++ Added dynamic port sizes to CrossEntropyLossNM
++ Changed CrossEntropyLoss to CrossEntropyLossNM to prevent confusion with pytorch's CrossEntropyLoss

File: examples/nlp/glue_benchmark/glue_benchmark_with_bert.py [hunk 1]
```diff
@@ -68,10 +68,10 @@
 import nemo.collections.nlp as nemo_nlp
 import nemo.core as nemo_core
 from nemo import logging
-from nemo.backends.pytorch.common import CrossEntropyLoss, MSELoss
+from nemo.backends.pytorch.common import CrossEntropyLossNM, MSELoss
 from nemo.collections.nlp.callbacks.glue_benchmark_callback import eval_epochs_done_callback, eval_iter_callback
 from nemo.collections.nlp.data import NemoBertTokenizer, SentencePieceTokenizer
-from nemo.collections.nlp.data.datasets.glue_benchmark_dataset import output_modes, processors
+from nemo.collections.nlp.data.datasets.glue_benchmark_dataset.glue_benchmark_dataset import output_modes, processors
```
Comment (okuchaiev) [inline examples/nlp/glue_benchmark/glue_benchmark_with_bert.py]: why is this name "glue_benchmark_dataset" repeated ?

File: nemo/backends/pytorch/common/losses.py [hunk 2]
```diff
@@ -47,7 +45,14 @@ def output_ports(self):
...
     def __init__(
-        self, pad_id=0, smoothing_coef=0.0, sample_wise=False, aux_ctc=False, ctc_initial_coef=0.1, ctc_blank_id=None
+        self,
+        pad_id=0,
+        smoothing_coef=0.0,
+        sample_wise=False,
+        aux_ctc=False,
+        ctc_initial_coef=0.1,
+        ctc_blank_id=None,
+        eps=1e-5,
```
Comment (okuchaiev) [inline nemo/backends/pytorch/common/losses.py]: eps is missing from docstring

File: nemo/backends/pytorch/common/losses.py [hunk 3]
```diff
@@ -95,10 +101,14 @@ def _ctc_loss(self, log_probs, targets, pad_mask):
         return loss
 
 
-class CrossEntropyLoss(LossNM):
+class CrossEntropyLossNM(LossNM):
     """
     CrossEntropyLoss
-
+    Args:
```
Comment (okuchaiev) [inline nemo/backends/pytorch/common/losses.py]: these args should correspond to __init__'s args

File: nemo/backends/pytorch/common/losses.py [hunk 4]
```diff
@@ -166,3 +187,51 @@ def __init__(self, reduction='mean'):
...
         loss = self._criterion(preds, labels)
         return loss
+
+
+class LossAggregatorNM(LossNM):
+    """
+    Neural module which combines sums several losses into one.
+
+    Args:
+        num_inputs (int): number of input losses
```
Comment (okuchaiev) [inline nemo/backends/pytorch/common/losses.py]: please document weights argument

Comment (VahidooX) [reply nemo/backends/pytorch/common/losses.py]: Added.

File: nemo/backends/pytorch/common/losses.py (hunk 2 above)
Comment (VahidooX) [reply nemo/backends/pytorch/common/losses.py]: added.

File: nemo/backends/pytorch/common/losses.py (hunk 3 above)
Comment (VahidooX) [reply nemo/backends/pytorch/common/losses.py]: updated.

File: examples/nlp/glue_benchmark/glue_benchmark_with_bert.py (hunk 1 above)
Comment (ekmb) [reply examples/nlp/glue_benchmark/glue_benchmark_with_bert.py]: one is the folder, another one is a file. removed the repetition

File: nemo/collections/nlp/nm/losses/masked_xentropy_loss.py [hunk 5]
```diff
@@ -42,16 +42,16 @@
 from nemo.core.neural_types import LabelsType, LengthsType, LogitsType, LossType, NeuralType
 from nemo.utils.decorators import add_port_docs
 
-__all__ = ['TRADEMaskedCrossEntropy', 'CrossEntropyLoss3D']
+__all__ = ['MaskedXEntropyLoss']
 
 
-class TRADEMaskedCrossEntropy(LossNM):
+class MaskedXEntropyLoss(LossNM):
```
Comment (blisc) [inline nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: Are you sure this should be a lossNM and not a Non-TrainableNM? I do not see any loss calculation in here

File: nemo/collections/nlp/nm/losses/smoothed_cross_entropy_loss.py [hunk 6]
```diff
@@ -16,10 +16,65 @@
...
+        predict_last_k (int): how many last tokens to use for the loss
+            calculation, important for fast evaluation of LM perplexity
+    """
+
+    @property
+    def input_ports(self):
+        """Returns definitions of module input ports.
+        """
+        return {
+            "logits": NeuralType(('B', 'T', 'D'), LogitsType()),
```
Comment (blisc) [inline nemo/collections/nlp/nm/losses/smoothed_cross_entropy_loss.py]: Shouldn't we rename this to log_probabilities?

File: nemo/collections/nlp/nm/losses/masked_xentropy_loss.py (hunk 5 above)
Comment (blisc) [inline nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: Can we consistent in using either XEntropy or CrossEntropy? I vote for CrossEntropy

File: nemo/collections/nlp/utils/functional_utils.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    total_sum = 0.0
+    for score in scores:
+        x = math.exp(score - max_score)
+        exp_scores.append(x)
+        total_sum += x
+
+    probs = []
+    for score in exp_scores:
+        probs.append(score / total_sum)
+    return probs
```
Comment (blisc) [inline nemo/collections/nlp/utils/functional_utils.py]: When would we ever want to do this without going through numpy or torch?

File: nemo/collections/nlp/nm/losses/masked_xentropy_loss.py (hunk 5 above)
Comment (VahidooX) [reply nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: It actually calculates log loss (binary cross entropy) for a selected element among the all labels. A faster and more over fitting robust version of cross entropy for multi class.

Comment (VahidooX) [reply nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: Changed it to MaskedLogLoss.

Comment (blisc) [reply nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: I see what you mean now. You are using labels to gather your logits.

Comment (blisc) [reply nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: Thanks for the clarification

Comment (blisc) [reply nemo/collections/nlp/nm/losses/masked_xentropy_loss.py]: why not MaskedCrossEntropyLoss?
